components/brave_perf_predictor/browser/perf_predictor_tab_helper_browsertest.cc

```diff
@@ -0,0 +1,156 @@
+
+  void InitEmbeddedTestServer() {
+    brave::RegisterPathProvider();
+    base::FilePath test_data_dir;
+    base::PathService::Get(brave::DIR_TEST_DATA, &test_data_dir);
+    embedded_test_server()->ServeFilesFromDirectory(test_data_dir);
+    content::SetupCrossSiteRedirector(embedded_test_server());
+    ASSERT_TRUE(embedded_test_server()->Start());
+  }
+
+  void WaitForAdBlockServiceThreads() {
+    scoped_refptr<base::ThreadTestHelper> tr_helper(new base::ThreadTestHelper(
+        g_brave_browser_process->local_data_files_service()->GetTaskRunner()));
+    ASSERT_TRUE(tr_helper->Run());
+    scoped_refptr<base::ThreadTestHelper> io_helper(new base::ThreadTestHelper(
+        base::CreateSingleThreadTaskRunner({content::BrowserThread::IO})
+            .get()));
+    ASSERT_TRUE(io_helper->Run());
+  }
+};
+
+IN_PROC_BROWSER_TEST_F(PerfPredictorTabHelperTest, NoBlockNoSavings) {
+  EXPECT_EQ(browser()->profile()->GetPrefs()->GetUint64(
+                brave_perf_predictor::prefs::kBandwidthSavedBytes),
+            0ULL);
+
+  GURL url = embedded_test_server()->GetURL("/blocking.html");
+  ui_test_utils::NavigateToURL(browser(), url);
+  content::WebContents* contents =
+      browser()->tab_strip_model()->GetActiveWebContents();
+
+  bool as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "setExpectations(1, 0, 0, 0, 0, 0);"
+                                          "addImage('logo.png')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+  // Prediction triggered when web contents are closed
+  contents->Close();
+  EXPECT_EQ(browser()->profile()->GetPrefs()->GetUint64(
+                brave_perf_predictor::prefs::kBandwidthSavedBytes),
+            0ULL);
+}
+
+IN_PROC_BROWSER_TEST_F(PerfPredictorTabHelperTest, ScriptBlockHasSavings) {
+  ASSERT_TRUE(g_brave_browser_process->ad_block_custom_filters_service()
+                  ->UpdateCustomFilters("*analytics.js"));
+  EXPECT_EQ(browser()->profile()->GetPrefs()->GetUint64(
+                brave_perf_predictor::prefs::kBandwidthSavedBytes),
+            0ULL);
+
+  GURL url = embedded_test_server()->GetURL("/blocking.html");
+  ui_test_utils::NavigateToURL(browser(), url);
+  content::WebContents* contents =
+      browser()->tab_strip_model()->GetActiveWebContents();
+
+  bool as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "setExpectations(0, 0, 0, 0, 1, 0);"
+                                          "xhr('analytics.js')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+  EXPECT_EQ(browser()->profile()->GetPrefs()->GetUint64(kAdsBlocked), 1ULL);
+  // Prediction triggered when web contents are closed
+  contents->Close();
+  EXPECT_NE(browser()->profile()->GetPrefs()->GetUint64(
+                brave_perf_predictor::prefs::kBandwidthSavedBytes),
+            0ULL);
```

Comment:
I'm a bit confused - there is no savings, but the test is named as `...HasSavings`

Reply:
ah, this is EXPECT_NE, everything is fine then :)